[rbi] Remove code that caused us to misidentify certain captured parameters as sending.

gottesmm opened the pull request:
Specifically, this code was added because otherwise we would in swift 5 + strict-concurrency mode emit two warnings, one at the AST level and one at the SIL level. Once we are in swift-6 mode, this does not happen since we stop compiling at the AST level since we will emit the AST level diagnostic as an error.

To do this, we tried to pattern match what the AST was erroring upon and treat the parameter as disconnected instead of being isolated. Sadly, this resulted in us treating certain closure cases incorrectly and not emit a diagnostic (creating a concurrency hole).

Given that this behavior results in a bad diagnostic only to avoid emitting two diagnostics in a mode which is not going to last forever... it really doesn't make sense to keep it. We really need a better way to handle these sorts of issues. Perhaps a special semantic parameter put on the function that squelches certain errors. But that is something for another day. The specific case it messes up is:

```
class NonSendable {
    func action() async {}
}

@MainActor
final class Foo {
    let value = NonSendable()

    func perform() {
        Task { [value] in
            await value.action() // Should emit error but do not.
        }
    }
}
```

In this case, we think that value is sending... when it isnt and we should emit an error.

rdar://146378329

gottesmm committed:
[rbi] Remove code that caused us to misidentify certain captured parameters as sending.

Specifically, this code was added because otherwise we would in swift 5 +
strict-concurrency mode emit two warnings, one at the AST level and one at the
SIL level. Once we are in swift-6 mode, this does not happen since we stop
compiling at the AST level since we will emit the AST level diagnostic as an
error.

To do this, we tried to pattern match what the AST was erroring upon and treat
the parameter as disconnected instead of being isolated. Sadly, this resulted in
us treating certain closure cases incorrectly and not emit a diagnostic
(creating a concurrency hole).

Given that this behavior results in a bad diagnostic only to avoid emitting two
diagnostics in a mode which is not going to last forever... it really doesn't
make sense to keep it. We really need a better way to handle these sorts of
issues. Perhaps a special semantic parameter put on the function that squelches
certain errors. But that is something for another day. The specific case it
messes up is:

```
class NonSendable {
    func action() async {}
}

@MainActor
final class Foo {
    let value = NonSendable()

    func perform() {
        Task { [value] in
            await value.action() // Should emit error but do not.
        }
    }
}
```

In this case, we think that value is sending... when it isnt and we should emit
an error.

rdar://146378329

jamieQ commented:
for visibility, i believe the examples in each of:

- https://github.com/swiftlang/swift/issues/76003
- https://github.com/swiftlang/swift/issues/76016
- https://github.com/swiftlang/swift/issues/78360

produce diagnostics with this change

gottesmm commented:
@swift-ci test windows platform

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
Looks like we hit an error in CoreLibsFoundation

gottesmm commented:
I found the problem. Reproducer:

```
class NS {}

@MainActor func transferToMain<T>(_ t: T) async {}

func test(_ x: NS) async {
  nonisolated(unsafe) let y = x
  await transferToMain(y)
  let _ = await {
    await transferToMain(y)
  }()
}
```

I haven't checked yet, but I am 995 sure the problem is that we do not understand that the variable in the closure is nonisolated(unsafe).

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
[rbi] Teach RBI how to infer nonisolated(unsafe) of closure captures.

The previous commit in this PR exposed that we were not handling this correctly.
Specifically, we incorrectly started to error in SwiftFoundation.

rdar://162629359

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

jamieQ commented on test/Concurrency/transfernonsendable_nonisolatedunsafe.swift:
```diff
@@ -1043,3 +1043,65 @@ func closureTests() async {
+
+  func testAsyncLet() async {
+    nonisolated(unsafe) let x = NonSendableKlass()
+    async let y: () = {
```
that makes sense, thanks. making a change like that probably would be more appropriate to do independently, but i was just curious.

gottesmm replied:
No worries! It is a natural question to ask!

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented on test/Concurrency/transfernonsendable.swift:
```diff
@@ -2001,3 +2000,22 @@ nonisolated(nonsending) func testCallNonisolatedNonsending(_ x: NonSendableKlass
   // expected-ni-ns-note @-2 {{sending task-isolated 'x' to @concurrent global function 'useValueAsyncConcurrent' risks causing data races between @concurrent and task-isolated uses}}
 }
+
+func avoidThinkingClosureParameterIsSending() {
```
Yes. I am aware of that. That isn't what was the issue. The issue was that there was an ABI issue preventing me from reproducing locally. I was curious if the fix I put in impacted the ability to make further progress.

gottesmm commented:
https://github.com/swiftlang/sourcekit-lsp/pull/2345

@swift-ci smoke test
